### PR TITLE
[JENKINS-30133] Workflow compatibility

### PR DESCRIPTION
**Pull request by amuniz**
Downstream of https://github.com/jenkinsci/analysis-core-plugin/pull/31

[JENKINS-30133](https://issues.jenkins-ci.org/browse/JENKINS-30133)

This changes allow to use ```CcmPublisher``` in a Workflow script, for example:

```
node {
  step([$class: 'CcmPublisher', pattern: '**/ccm.xml', failedTotalAll: '0', usePreviousBuildAsReference: false])
}
```

### CODE REVIEW

File: src/main/java/hudson/plugins/ccm/CcmResult.java [hunk 1]
```diff
@@ -44,8 +44,9 @@
 	 * @param defaultEncoding
 	 * @param result
 	 */
-	public CcmResult(final AbstractBuild<?, ?> build, final String defaultEncoding, final ParserResult result) {
-		super(build, defaultEncoding, result, new BuildHistory(build, CcmResultAction.class));
+	public CcmResult(final Run<?, ?> build, final String defaultEncoding, final ParserResult result) {
+		super(build, new BuildHistory(build, CcmResultAction.class, false, false), result, defaultEncoding);
+		serializeAnnotations(result.getAnnotations());
```
Comment (amuniz) [inline src/main/java/hudson/plugins/ccm/CcmResult.java]: This change is not related to workflow, but it was needed after upgrading to analysis-core upstream snapshot.
See [this change](https://github.com/jenkinsci/analysis-core-plugin/commit/3185a5e28a16b0088b370a2edb6fe145a80dda46) which requires a explicit call to `serializeAnnotations(result.getAnnotations())`

**Comment by amuniz** [comment]
@reviewbybees

**Comment by jglick** [comment]
:bee:

**Comment by recena** [comment]
:bee:

File: src/main/java/hudson/plugins/ccm/CcmPublisher.java [hunk 2]
```diff
@@ -57,26 +59,13 @@
...
-                unstableTotalAll, unstableTotalHigh, unstableTotalNormal, unstableTotalLow,
-                unstableNewAll, unstableNewHigh, unstableNewNormal, unstableNewLow,
-                failedTotalAll, failedTotalHigh, failedTotalNormal, failedTotalLow,
-                failedNewAll, failedNewHigh, failedNewNormal, failedNewLow,
-                canRunOnFailed, shouldDetectModules, canComputeNew, PLUGIN_NAME);
-        this.pattern = pattern;
+    public CcmPublisher() {
+        super(PLUGIN_NAME);
     }
-
```
Comment (recena) [inline src/main/java/hudson/plugins/ccm/CcmPublisher.java]: @amuniz Unrelated change? jajajaja

Comment (amuniz) [inline src/main/java/hudson/plugins/ccm/CcmPublisher.java]: :smile:

**Comment by amuniz** [comment]
@reviewbybees done

**Comment by amuniz** [comment]
@kinow please, consider this PR for merge if you think everything is fine. Thanks.

**Comment by amuniz** [comment]
@kinow upstream [Analysis Core was released](https://github.com/jenkinsci/analysis-core-plugin/releases/tag/analysis-core-1.73) some days ago. This PR is un-blocked now. Do you think it could be merged? Thanks!

**Comment by kinow** [comment]
Just to let you know that I haven't forgotten about this PR. Just busy with another Jenkins related development cycle :-) Will try to review/merge it in the next days. Sorry for the delay.

**Comment by amuniz** [comment]
Great @kinow! Thanks for your work.

File: pom.xml [hunk 3]
```diff
@@ -67,14 +69,31 @@
         <dependency>
             <groupId>org.jvnet.hudson.plugins</groupId>
             <artifactId>analysis-core</artifactId>
-            <version>1.71</version>
-            <type>hpi</type>
+            <version>1.73-beta-SNAPSHOT</version><!-- TODO: change to 1.73 before release -->
```
Comment (kinow) [inline pom.xml]: I believe 1.73 has been released (see https://github.com/jenkinsci/analysis-core-plugin/tree/analysis-core-1.73). 

Can I change it to 1.73 after merging? 

ps: sorry, it took me so long to review, that actually 1.74 is out as well.

File: src/test/java/hudson/plugins/ccm/workflow/WorkflowCompatibilityTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import hudson.model.Result;
+import hudson.plugins.ccm.CcmResultAction;
+
+import static org.junit.Assert.*;
+
+/**
+ * Test workflow compatibility.
+ */
+public class WorkflowCompatibilityTest {
+
```
Comment (kinow) [inline src/test/java/hudson/plugins/ccm/workflow/WorkflowCompatibilityTest.java]: Great! :D Thanks for adding tests.

File: src/main/java/hudson/plugins/ccm/CcmResult.java (hunk 1 above)
Comment (kinow) [inline src/main/java/hudson/plugins/ccm/CcmResult.java]: Thanks for the heads up! :+1:

File: src/main/java/hudson/plugins/ccm/CcmPublisher.java (hunk 2 above)
Comment (kinow) [inline src/main/java/hudson/plugins/ccm/CcmPublisher.java]: Wut? haha, haven't looked at the analysis core, and to be honest the CCM migration to the analysis was a copy+paste from another plug-in.

Do you reckon it won't break existing jobs?

File: pom.xml (hunk 3 above)
Comment (amuniz) [inline pom.xml]: Yes, you can safely change it to 1.73.

File: src/main/java/hudson/plugins/ccm/CcmPublisher.java (hunk 2 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/ccm/CcmPublisher.java]: It should not. I took care of backward compatibility, but if you see something weird, tell me!

NOTE: the initial comment of this thread was a kind of joke from my teammate, you can ignore it.

File: pom.xml (hunk 3 above)
Comment (kinow) [inline pom.xml]: Thanks @amuniz ! Here's the test steps that I followed:

1. Created a simple job, using the XML files in src/test/resources
2. Tested that it was working correctly, and the graph was generated OK as well
3. Checked out pr/5 branch locally
4. Executed the job again
5. Everything worked OK, except the graph, that was empty
6. Updated to 1.73 the analysis-core. Same error.
7. Updated to 1.74, and now it worked!

Merging your pull request, but will bump analysis-core to 1.74.

Thanks and sorry for the delay to review and merge it.

Cheers,
Bruno

**Comment by kinow** [comment]
Releasing in 3.1

File: pom.xml (hunk 3 above)
Comment (amuniz) [inline pom.xml]: It seems that a bug in trend graphs were fixed in analysis-core 1.74. I think it's safe to upgrade to 1.74 since there are no big changes (only a bug fix).
Thanks Bruno!

Comment (kinow) [inline pom.xml]: Didn't have time to check it, but thanks for confirming there was an issue in 1.73.

3.1 has been released about 45 minutes ago. Should be available through the update centre in the next hours.

Comment (amuniz) [inline pom.xml]: Great!

FYI [the fix in analysis-core](https://github.com/jenkinsci/analysis-core-plugin/commit/34c2aef5018ec716c7a6b6ba160eea429bcd7755)
